### DISCUSSION TITLE
How to gather results from different process

### DISCUSSION THREAD

**Discussion by mrbeann** [hpcaitech/ColossalAI#2044] [Community | Q&A]
I try to train a model with ColossalAI and want to calculate some metrics like accuracy. Thus, I record the predictions and truth as follows. But the results show it only contains single batch results. What should I do to get the correct results? (maybe something like `torch.distributed.all_gather` is needed?)
```python
predictions = []
truths = []
for index, (data_v, labels_v) in enumerate(val_dl):
    data_v, labels_v = data_v.cuda(), labels_v.cuda()
    with torch.no_grad():
        logits = engine(data_v)
    loss = engine.criterion(logits, labels_v)
    running_loss += loss.item()
    final_prob = nn.Softmax(dim=-1)(logits)
    final = final_prob.argmax(dim=-1)
    predictions.append(final)
    truths.append(labels_v)
predictions = np.array(predictions.cpu())
truths = np.array(truths.cpu())
print(predictions.shape, truths.shape, len(val_dl))
```

**Comment by feifeibear** [comment]
Thanks for your feedback. I believe you should define your model in a torch.nn.Module class first. The code snippet is not a typical PyTorch Model.

**Reply by mrbeann** [reply]
Yeah, I have a model in torch.nn.Module class (similar to Resnet), and transfer it to `engine` with `colossalai.initialize`

**Reply by feifeibear** [reply]
I see. Looks like you are using an old API and capsule model, optimizers, LR schedulers, etc in an engine.
Using Engine is a high-level API, you need to use some hooks to fetch accuracy metrics.
@FrankLeeeee can you help to answer his issue?

**Reply by mrbeann** [reply]
Yeah, I was unable to find examples that can achieve this. It will be great to have some examples of the hooks.

**Reply by mrbeann** [reply]
> I see. Looks like you are using an old API and capsule model, optimizers, LR schedulers, etc in an engine. Using Engine is a high-level API, you need to use some hooks to fetch accuracy metrics. @FrankLeeeee can you help to answer his issue?

BTW, I'm new to colossalAI. According to the documentation, it seems Engine is more flexible than Trainer. Is there any new and low-level API?

**Reply by feifeibear** [reply]
You can simply use for loop with fwd/bwd/step for training (similar to PyTorch usage)
see our vit example.
https://github.com/hpcaitech/ColossalAI/blob/main/examples/images/vit/train.py#L95

**Reply by mrbeann** [reply]
Can this API deal with the sync automatically? I do not find any code related to it.

**Reply by feifeibear** [reply]
What do you mean sync automatically?

**Reply by mrbeann** [reply]
E.g., in 
https://github.com/hpcaitech/ColossalAI/blob/main/examples/images/vit/train.py#L143
if I use a list to record the output.

```
res = []
for index, (x, y) in tqdm(enumerate(test_dataloader), total=len(test_dataloader), leave=False):
    x, y = x.cuda(), y.cuda()
    output = model(x)
    res.append(output)
    test_loss += F.cross_entropy(output, y, reduction='sum').item()
    pred = output.argmax(dim=1, keepdim=True)
    correct += pred.eq(y.view_as(pred)).sum().item()
    test_sum += y.size(0)
print(res)
```

Will this work properly?

**Reply by feifeibear** [reply]
It's supposed to work properly if you update the following line. Since the output may be a shard tensor, we need to convert it to replicate, so its shape is as you expected.
```
res.append(output.to_replicate())
```

**Reply by mrbeann** [reply]
It seems this API uses `ColoTensor`, so I need another dataloader that also uses ColoTensor?
